RFC to introduce Cloud Native Buildpacks lifecycle

modulo11 opened the pull request:
This RFC tries to gently introduce Cloud Native Buildpacks to Cloud Foundry. It builds upon [RFC 0017](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md) but introduces CNBs directly in the platform instead of shimming them.

[Preview](https://github.com/sap-contributions/cf-community/blob/cnb-lifecycle/toc/rfc/rfc-draft-cnb-lifecycle.md)

@cloudfoundry/toc 
CC @loewenstein @c0d1ngm0nk3y, @pbusko, @nicolasbender, @beyhan, @stephanme, @stefanlay, @PlamenDoychev

FloThinksPi commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,135 @@
+
+#### Add Paketo Stack(s)
+
+This RFC does not include the addition of a new stack to Cloud Foundry, rather the resulting droplet would run on top of the existing `cflinuxfs4`. This should work for most, if not all apps, as the stack is much bigger (in terms of native packages installed) than the stacks used by Paketo.
```
The mechanism that currently exists rebases the base layer(stack) on every application restart onto the droplet.
The behaviour CNBs have is to add the base image (stack equivalent) at build time to a resulting OCI Image.
The way this mechanism works is different when building internally or externally with the same CNB.

The way CNBs add the base layer while building can be favorable for the stack depreciation problem where currently its not possible to remove a stack with application still depending on it. When the stack gets added at build time this changes as then missing stacks just prevent restaging. Already build droplets include the stack in the droplet and keep running.
However this also comes with a some downsides:
- Stack update means restaging all apps(resource costly, complex)
- The stack is saved multiple times in the blobstore(storage and traffic)

Although using the build workflow that is common for CNBs in CF would help plus make the systems behave similar it comes with some costs im not sure if they are worth it yet but maybe something to consider.

loewenstein replied:
I'd say that the thought of using CNBs to produce full OCI images is independent from this "potential future work". Also for a droplet based approach, the smaller Paketo base images in *Static*, *Tiny* and *Base* are an improvement for application security imho.

loewenstein replied:
@FloThinksPi I would say you are calling for a separate RFC that's largely independent from the Cloud Native Buildpack topic. Are we good to resolve this question?

alexbarbato replied:
Not to de-rail - but is this a point worth considering further:

> The stack is saved multiple times in the blobstore(storage and traffic)
This is a core benefit of an OCI registry that it should see the layers and thus save a tremendous amount of space. Re-using the blobstore certainly is simplest, but there are cons on the other end to my eye.

Maybe that concern belongs in the follow on RFC?

loewenstein-sap replied:
I am not sure if moving CF integrated CNBs from creating droplets to creating container images is something we want. If we don't, then there is no problem with a stack saved multiple times, because the roots is saved separately from the droplets.

stephanme commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,135 @@
+
+### Alternative APIs
+
+- Instead of a binary switch, introduce a `buildpack-type` (`v2`/`v3` or `cf`/`cnb` or `classic`/`cnb`) to distinguish between different lifecycles
```
+1 for `buildpack-type` in app manifest. It allows smooth extension in the future while a `cnb: true` means that every new lifecycle type needs additional properties. Future extensions might be a CNB lifecycle that builds OCI images (including stack) instead of traditional droplets (as @FloThinksPi suggested below).

`buildpack-type` matches to a certain extent the `lifecycle.type` attribute in the [CF API](https://v3-apidocs.cloudfoundry.org/index.html#lifecycles). Full consistency between CF API, app manifest and cf cli is probably out of reach due to history.

loewenstein replied:
As you explicitly mention a potential CNB based lifecycle that introduces full OCI images, we should then probably not use `cnb`, but rather something like `cnb-droplet`. WDYT?

PlamenDoychev replied:
I would also prefer to have `buildpack-type` instead of cnb: true, which more or less looks like a switch.

pbusko replied:
We've updated the RFC and changed the `cnb: true|false` to the `lifecycle: buildpack|docker|cnb`. However, the introduction of `cnb-oci` or `cnb-droplet` seems to be too early. What would be the benefit for the end user to be able to choose between droplet-style and raw OCI? But since the whole feature will be behind a feature flag, the behaviour could be also controller by administrators on the CCNG config level.

stephanme replied:
Potential end user benefit:
- cnb-droplet: user gets stack updates automatically, similar to buildpack lifecycle
- cnb-oci: user gets a standard OCI image that could be used outside outside of CF as well (ok, a rather artificial benefit as there are other options to build such an image), but no automatic stack updates (need a restage) 

So yes, one cnb lifecycle should be sufficient for today.

dsboulder commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,135 @@
+
+## Problem
+
+The v2 buildpacks are effectively in maintenance mode and do not receive substantial new features. By not integrating with v3 buildpacks, Cloud Foundry is missing out on new buildpacks (e.g. Java Native and Web Servers CNBs) as well as any new features that are added to the still-actively-developed v3 buildpacks.
```
The folks on our side are doing development on both V2 and CNBs in parallel, I'm seeing new capabilities come online for V2 frequently as well. What's giving you folks this impression? This will be significant work in quite a few areas to support CNBs in CF and I want to make it's worth it.

loewenstein replied:
We've taken this pretty literally from [RFC 0017](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md#problem).

Regarding the amount of work...
> This will be significant work in quite a few areas to support CNBs in CF and I want to make it's worth it.

... our plan was not only to propose this RFC, but also to work on contributing the implementation.

loewenstein replied:
@dsboulder would you want to propose a different wording or are we good to resolve this question?

dsboulder commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,135 @@
+  - No OCI registry necessary
+  - Reuse cflinuxfs4 as rootfs during build and run
+
+### High Level Implementation
```
I never saw a _great_ solution to the VCAP services => CN bindings problem? Is there a way to make this work that isn't a ton of brittle translation tables?

beyhan replied:
The https://github.com/cloudfoundry/community/pull/804 tries to address this point. I would be interested in your feedback. 

FYI @dmikusa as we discussed this during the TOC meeting and feedback is welcome.

loewenstein replied:
@rkoster @dsboulder @beyhan Should we add a little subsection referring to https://github.com/buildpacks/libcnb/pull/228, the possibility to do the same for `packit` based Paketo buildpacks and referring to #804 for solving it once and for all?

In short, I would probably write this as
> Buildpacks that need credentials from services should be adapted to read from `VCAP_SERVICES` until #804 removes this requirement.
> **Note**: Buildpacks based on `libpak` or `libcnb` already understand credentials from `VCAP_SERVICES` since [libcnb#228](https://github.com/buildpacks/libcnb/pull/228) has been merged and released.

c0d1ngm0nk3y replied:
> Ideally a heading would be added discussion what to do with service bindings.

Should it really? It feels a bit out of place because...

a) This is not necessarily `cf` related, but rather buildpack related
b) This feels like an implementation detail as already established in the predecessor rfc ([here](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md#service-bindings))

What should we mention here? That `cf` does not have to change?

loewenstein replied:
@beyhan like https://github.com/cloudfoundry/community/pull/796/files#r1574769369?

beyhan replied:
yes, thanks!

rkoster replied:
> Should it really? It feels a bit out of place because...
Sure now that we have an RFC addressing the problem of servicebinding spec compatibility there is no need to discuss it in this RFC. A link to the service-binding RFC would be nice so that future readers understand how they relate.

modulo11 commented:
@beyhan we added a section regarding consuming buildpacks from private registries. We did not yet decide on an implementation. Does this require additional reviewers?

beyhan commented:
> @beyhan we added a section regarding consuming buildpacks from private registries. We did not yet decide on an implementation. Does this require additional reviewers?

When the discussions are resolved I will require review from the involved people any way. Regarding the change could you please commend why the current support for private registries described in https://docs.cloudfoundry.org/devguide/deploy-apps/push-docker.html#registry isn't helpful. I think this comment/explanation could be added also to the RFC

modulo11 commented:
> > @beyhan we added a section regarding consuming buildpacks from private registries. We did not yet decide on an implementation. Does this require additional reviewers?
> 
> @modulo11 When the discussions are resolved I will require review from the involved people anyway. Regarding the change could you please comment why the current support for private registries described in https://docs.cloudfoundry.org/devguide/deploy-apps/push-docker.html#registry isn't helpful or can't be reused. I think this comment/explanation could be added also to the RFC.

I updated the section. It's basically the lack of support for multiple registries.

loewenstein commented:
@cloudfoundry/toc if @rkoster and @beyhan are ok with https://github.com/cloudfoundry/community/pull/796#discussion_r1570898862, i.e. relying on the [Service Binding section in the older CNB RFC](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0017-add-cnbs.md#service-bindings), there are no pending discussions left I believe.

How do we bring this forward for a vote? I guess it would be good to have explicit buy-in to review the necessary changes in @cloudfoundry/wg-app-runtime-platform (new app lifecycle in the Diego Bosh release) and @cloudfoundry/wg-app-runtime-interfaces (CAPI and CLI changes). Once the vote is done and positive and this RFC merged, we'd need to request the creation of one new repository (for the new app lifecycle) - is there a process for this?

beyhan commented:
@loewenstein I added your last comment as topic for the next TOC meeting.

loewenstein commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,176 @@
+- No fundamental changes to the architecture of CF
+  - Result of the staging process will be a droplet
+  - No OCI registry necessary
+  - Reuse cflinuxfs4 as rootfs during build and run
```
```suggestion
  - Reuse cflinuxfs4 as rootfs during build and run
  - No change to how the Cloud Foundry platform provides service binding information
```

beyhan commented:
We discussed the last commend during the TOC meeting today 23.04.24 and we decided to start the FCP with the goal to approve this RFC next Tuesday 30.04.24.

natalieparellano commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,177 @@
+- Increase adoption of Cloud Native Buildpacks
+- Open the door for eventual deprecation of the v2 buildpacks, reducing maintenance costs (v2 buildpack deprecation is NOT included in this RFC)
+- No fundamental changes to the architecture of CF
+  - Result of the staging process will be a droplet
```
The CNB lifecycle supports exporting to OCI layout on disk (the resulting image can even be "sparse" in that run image layers will be missing). This should facilitate exporting to a CF droplet without having to re-implement large parts of the exporter.

beyhan commented:
@loewenstein  Additionally, we discussed that @ameowlia, @rkoster and @stephanme will support the contributions to get approved. You can also open a pr like https://github.com/cloudfoundry/community/pull/730 to add the new lifecycle repository and @ameowlia will help with finding the right place for it.

loewenstein commented on toc/rfc/rfc-draft-cnb-lifecycle.md:
```diff
@@ -0,0 +1,177 @@
+  - Result of the staging process will be a droplet
+  - No OCI registry necessary
+  - Reuse cflinuxfs4 as rootfs during build and run
+  - No change to how the Cloud Foundry platform provides service binding information
```
```suggestion
  - No change to how the Cloud Foundry platform provides service binding information (see [#804](https://github.com/cloudfoundry/community/pull/804) for how this will be covered in a separate RFC)
```

loewenstein replied:
@beyhan @rkoster ^^
